### PR TITLE
Inertialnav removal

### PR DESCRIPTION
**Pull request by khancyr**
based on https://github.com/ArduPilot/ardupilot/pull/6718

I am open to rename new function like get_relative_position_NEU_origin_cm, or try to split more the PR for the rename stuff

### CODE REVIEW

File: ArduCopter/commands.cpp [hunk 1]
```diff
@@ -29,7 +29,8 @@ void Copter::set_home_to_current_location_inflight() {
     // get current location from EKF
     Location temp_loc;
     if (ahrs.get_location(temp_loc)) {
-        const struct Location &ekf_origin = inertial_nav.get_origin();
+        struct Location ekf_origin;
+        ahrs.get_origin(ekf_origin);
```
Comment (rmackay9) [inline ArduCopter/commands.cpp]: We should probably check the return value of ahrs.get_origin()

File: ArduCopter/commands.cpp [hunk 2]
```diff
@@ -106,7 +107,8 @@ bool Copter::set_home(const Location& loc, bool lock)
 bool Copter::far_from_EKF_origin(const Location& loc)
 {
     // check distance to EKF origin
-    const struct Location &ekf_origin = inertial_nav.get_origin();
+    struct Location ekf_origin;
+    ahrs.get_origin(ekf_origin);
```
Comment (rmackay9) [inline ArduCopter/commands.cpp]: We should check the return of ahrs.get_origin here as well.  if we can't get it we should return false.

File: ArduCopter/position_vector.cpp [hunk 3]
```diff
@@ -10,22 +10,25 @@
 // pv_location_to_vector - convert lat/lon coordinates to a position vector
 Vector3f Copter::pv_location_to_vector(const Location& loc)
 {
-    const struct Location &origin = inertial_nav.get_origin();
+    struct Location origin;
```
Comment (rmackay9) [inline ArduCopter/position_vector.cpp]: if the ahrs.get_origin() fails here (which will likely never happen) then we will use the "origin" value as-is.  It hasn't been explicitly initialised but I suspect we always initialise new variables to zero so it's probably ok.

File: ArduPlane/quadplane.cpp [hunk 4]
```diff
@@ -1824,7 +1824,8 @@ void QuadPlane::vtol_position_controller(void)
 void QuadPlane::setup_target_position(void)
 {
     const Location &loc = plane.next_WP_loc;
-    Location origin = inertial_nav.get_origin();
+    Location origin;
+    ahrs.get_origin(origin);
```
Comment (rmackay9) [inline ArduPlane/quadplane.cpp]: These quadplane.cpp changes should go in a commit that is separate from the Copter changes.

File: ArduSub/commands.cpp [hunk 5]
```diff
@@ -30,7 +30,8 @@ void Sub::set_home_to_current_location_inflight()
     // get current location from EKF
     Location temp_loc;
     if (ahrs.get_location(temp_loc)) {
-        const struct Location &ekf_origin = inertial_nav.get_origin();
+        struct Location ekf_origin;
```
Comment (rmackay9) [inline ArduSub/commands.cpp]: These sub changes should also be separated out into a separate commit.

File: ArduSub/commands.cpp [hunk 6]
```diff
@@ -102,7 +103,8 @@ bool Sub::set_home(const Location& loc, bool lock)
 bool Sub::far_from_EKF_origin(const Location& loc)
 {
     // check distance to EKF origin
-    const struct Location &ekf_origin = inertial_nav.get_origin();
+    struct Location ekf_origin;
+    ahrs.get_origin(ekf_origin);
```
Comment (rmackay9) [inline ArduSub/commands.cpp]: We should check the ahrs.get_origin() method's return value here too.

File: libraries/AP_InertialNav/AP_InertialNav.h [hunk 7]
```diff
@@ -40,13 +40,6 @@ class AP_InertialNav
      */
     virtual nav_filter_status get_filter_status() const = 0;
```
Comment (rmackay9) [inline libraries/AP_InertialNav/AP_InertialNav.h]: Changes to libraries should be separated out into their own commit for each library.

File: ArduCopter/AP_Arming.cpp [hunk 8]
```diff
@@ -495,7 +496,8 @@ bool AP_Arming_Copter::gps_checks(bool display_failure)
 bool AP_Arming_Copter::pre_arm_ekf_attitude_check()
 {
     // get ekf filter status
-    nav_filter_status filt_status = _inav.get_filter_status();
+    nav_filter_status filt_status;
+    _ahrs_navekf.get_filter_status(filt_status);
```
Comment (rmackay9) [inline ArduCopter/AP_Arming.cpp]: We should check the get-filter-status has returned true.  if it hasn't we should fail this check.  This applies to all other uses of this method as well.

File: ArduSub/inertia.cpp [hunk 9]
```diff
@@ -11,7 +11,9 @@ void Sub::read_inertia()
     current_loc.lat = inertial_nav.get_latitude();
 
     // exit immediately if we do not have an altitude estimate
-    if (!inertial_nav.get_filter_status().flags.vert_pos) {
+    nav_filter_status filt_status;
```
Comment (rmackay9) [inline ArduSub/inertia.cpp]: These sub changes should be separated out into a separate commit.  This applies to various other commits as well so I won't keep repeating this comment.

File: libraries/AC_PrecLand/AC_PrecLand.cpp [hunk 10]
```diff
@@ -97,7 +97,7 @@ const AP_Param::GroupInfo AC_PrecLand::var_info[] = {
 // Note that the Vector/Matrix constructors already implicitly zero
 // their values.
 //
-AC_PrecLand::AC_PrecLand(const AP_AHRS& ahrs, const AP_InertialNav& inav) :
+AC_PrecLand::AC_PrecLand(const AP_AHRS_NavEKF& ahrs, const AP_InertialNav& inav) :
```
Comment (rmackay9) [inline libraries/AC_PrecLand/AC_PrecLand.cpp]: We should remove AC_PrecLand's reference to inertialNav.  Perhaps that is done in a later commit.. in which case that other commit can be squashed into this one.

File: libraries/AP_InertialNav/AP_InertialNav_NavEKF.cpp [hunk 11]
```diff
@@ -78,20 +78,6 @@ float AP_InertialNav_NavEKF::get_altitude() const
 }
 
 /**
- * getHgtAboveGnd - get latest height above ground level estimate in cm and a validity flag
```
Comment (rmackay9) [inline libraries/AP_InertialNav/AP_InertialNav_NavEKF.cpp]: Instead of removing parts of inertial-nav one by one, perhaps just a single commit at the end to completely remove the library would be best.

File: ArduCopter/Copter.h [hunk 12]
```diff
@@ -471,6 +471,7 @@ class Copter : public AP_HAL::HAL::Callbacks {
     // 3D Location vectors
     // Current location of the copter (altitude is relative to home)
     Location_Class current_loc;
+    Location ekf_origin;
```
Comment (rmackay9) [inline ArduCopter/Copter.h]: I'm not sure it's a great idea to keep a copy of this around.  I know it's a bit more code and checking.. but better to just get it straight from the ekf each time.

File: libraries/AP_InertialNav/AP_InertialNav_NavEKF.cpp [hunk 13]
```diff
@@ -14,34 +14,23 @@
...
 
 /**
  * get_position - returns the current position relative to the home location in cm.
  *
  * @return
  */
-const Vector3f &AP_InertialNav_NavEKF::get_position(void) const 
+const Vector3f &AP_InertialNav_NavEKF::get_position(void) const
 {
+    // get the NED position relative to the local earth frame origin
```
Comment (rmackay9) [inline libraries/AP_InertialNav/AP_InertialNav_NavEKF.cpp]: I think we want to get rid of all these calls to AP_InertialNav so let's not spend time changing the internals of AP_InertialNav.

File: ArduCopter/system.cpp [hunk 14]
```diff
@@ -645,21 +645,21 @@ void Copter::allocate_motors(void)
     }
     AP_Param::load_object_from_eeprom(attitude_control, ac_var_info);
         
-    pos_control = new AC_PosControl(*ahrs_view, inertial_nav, *motors, *attitude_control,
+    pos_control = new AC_PosControl(*ahrs_view, ahrs, *motors, *attitude_control,
```
Comment (rmackay9) [inline ArduCopter/system.cpp]: If we are going to change the constructor for AC_PosControl, let's make the argument order consistent with AC_WPNav (i.e. ahrs before ahrs_view).

File: ArduCopter/Copter.h [hunk 15]
```diff
@@ -472,6 +472,7 @@ class Copter : public AP_HAL::HAL::Callbacks {
     // Current location of the copter (altitude is relative to home)
     Location_Class current_loc;
     Location ekf_origin;
+    Vector3f current_pos;
```
Comment (rmackay9) [inline ArduCopter/Copter.h]: Let's consider carefully if we want to maintain this current_pos variable which is just a copy of what's in the EKF.  Maybe we should get input from @peterbarker and @OXINARF.

File: ArduCopter/Copter.h [hunk 16]
```diff
@@ -473,6 +473,7 @@ class Copter : public AP_HAL::HAL::Callbacks {
     Location_Class current_loc;
     Location ekf_origin;
     Vector3f current_pos;
+    Vector3f current_vel;
```
Comment (rmackay9) [inline ArduCopter/Copter.h]: Re adding current_vel, one small thing is that it adds an additional dependency from the mode code to the main vehicle code which makes it harder to separate the mode code from the vehicle.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp [hunk 17]
```diff
@@ -407,7 +407,8 @@ void AC_PosControl::pos_to_rate_z()
 // calculates desired acceleration and calls accel throttle controller
 void AC_PosControl::rate_to_accel_z()
 {
-    const Vector3f& curr_vel = _ahrs.get_velocity();
+    Vector3f curr_vel;
```
Comment (rmackay9) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp]: any reason we're changing from _ahrs.get_velocity to _ahrs.get_velocity_NEU_cm?

**Comment by rmackay9** [comment]
In general it's looking pretty good, thanks very much for doing this!

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 17 above)
Comment (khancyr) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp]: it was _inav.get_velocity() to should be get_velocity_NEU_cm(), it should merge the renaming with the changes ... the flow will be better

File: ArduCopter/Copter.h (hunk 15 above)
Comment (khancyr) [inline ArduCopter/Copter.h]: The problem with those data is that their aren't exactly what is in EKF and need to allocate either a structur or a float vect to be get ... Plus by updating them only once we reach the inertia callback, it garanties that the behaviour remain the same than with inertialNav that was making a copy too.
But I am not blocked on that!

File: libraries/AC_PrecLand/AC_PrecLand.cpp (hunk 10 above)
Comment (khancyr) [inline libraries/AC_PrecLand/AC_PrecLand.cpp]: yep it is done later ... I will redone the external library change as one commit per library

File: ArduCopter/AP_Arming.cpp (hunk 8 above)
Comment (khancyr) [inline ArduCopter/AP_Arming.cpp]: ok!

**Comment by khancyr** [comment]
I will add the return check for get_origin etc !

**Comment by khancyr** [comment]
@rmackay9 Most call from ahrs.get_origin(ekf_origin) were removed with the addition of the ekf_origin parameter. Should I check the return in read_inertia() ?
checking the return type for get_filter_status() change the API a little as previously it use the 0 intialized value

File: libraries/AP_AHRS/AP_AHRS_NavEKF.h [hunk 18]
```diff
@@ -140,6 +140,7 @@ class AP_AHRS_NavEKF : public AP_AHRS_DCM
     // return true if the estimate is valid
     bool get_relative_position_NED_home(Vector3f &vec) const override;
     bool get_relative_position_NED_origin(Vector3f &vec) const override;
+    void get_relative_position_NEU_origin_cm(Vector3f &vec) const;
```
Comment (khancyr) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.h]: I agree that it isn't the best, but without it will be prone to error as all calculus on copter are made on NEU and in cm. So easy to forget it. The right direction should be to pass the inertialnav removal and then convert copter to NED.

File: libraries/AP_AHRS/AP_AHRS_NavEKF.h [hunk 19]
```diff
@@ -142,6 +142,7 @@ class AP_AHRS_NavEKF : public AP_AHRS_DCM
     bool get_relative_position_NED_home(Vector3f &vec) const override;
     bool get_relative_position_NED_origin(Vector3f &vec) const override;
     void get_relative_position_NEU_origin_cm(Vector3f &vec) const;
+    float get_altitude() const;
```
Comment (khancyr) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.h]: yep agree

**Comment by khancyr** [comment]
@rmackay9 @OXINARF would it mind you to reopen https://github.com/ArduPilot/ardupilot/pull/6718, so we can have easier to review and modify PR ? and the 5 last can already be merged safely !

File: libraries/AP_AHRS/AP_AHRS_NavEKF.h (hunk 18 above)
Comment (OXINARF) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.h]: I actually think that NEU is easier to work with. So I'm not against adding methods to return values in that coordinate system, but if we do it, we should add those for all methods, not just one.

Regarding centimeters, I disagree with you - having one method be different than all others is more error prone. Besides making the API inconsistent - we need to improve our library APIs, not make them worse. Returning meters also goes in the direction of using SI base units where possible and converting it to another unit where it is needed.
I'll also mention that if some place needs meters, we would be doing `* 100 / 100`, which does more operations than needed.

**Comment by OXINARF** [comment]
@khancyr Why is it easier to review #6718? It looks like at least 4 commits shouldn't be there? I think there are three choices here:

- one PR doing all changes
- multiple PRs, each changing a method from AP_InertialNav to something else, and a last one removing the entire library
- multiple PRs, each changing a subsystem from using AP_InertialNav methods to something else, and a last one removing the entire library

I'm fine with any of them - but for example, for Copter, I think we have some decision to make that should be the same for a couple of method changes and so option 2 will make it harder to make decisions.

File: libraries/AC_Avoidance/AC_Avoid.cpp [hunk 20]
```diff
@@ -397,7 +397,9 @@ Vector2f AC_Avoid::get_position() const
 {
     const Vector3f position_xyz = _inav.get_position();
     const Vector2f position_xy(position_xyz.x,position_xyz.y);
-    const Vector2f diff = location_diff(_inav.get_origin(),_ahrs.get_home()) * 100.0f;
+    Location origin;
+    _ahrs.get_origin(origin);
+    const Vector2f diff = location_diff(origin,_ahrs.get_home()) * 100.0f;
```
Comment (khancyr) [inline libraries/AC_Avoidance/AC_Avoid.cpp]: yep it is just get_relative_position_NE_home()*100.0f

File: libraries/AC_Avoidance/AC_Avoid.cpp [hunk 21]
```diff
@@ -407,7 +409,9 @@ Vector2f AC_Avoid::get_position() const
 float AC_Avoid::get_alt_above_home() const
 {
     // vehicle's alt above ekf origin + ekf origin's alt above sea level - home's alt above sea level
-    return _inav.get_altitude() + _inav.get_origin().alt - _ahrs.get_home().alt;
+    Location origin;
+    _ahrs.get_origin(origin);
+    return _inav.get_altitude() + origin.alt - _ahrs.get_home().alt;
```
Comment (khancyr) [inline libraries/AC_Avoidance/AC_Avoid.cpp]: mutilply by -100 as we are in cm !

**Comment by khancyr** [comment]
ok that should be better ! Still need to rework the get_altitude(), maybe add a better way to get z velocity alone.
And there was a bug in WPNav I think : https://github.com/ArduPilot/ardupilot/blob/master/libraries/AC_WPNav/AC_WPNav.cpp#L352, it set set_speed_xy with lenght on xyz vector, doesn't look right !

**Comment by khancyr** [comment]
OK rebased, and most comment addressed !

**Comment by khancyr** [comment]
@rmackay9 @OXINARF ping!

File: libraries/AC_WPNav/AC_Circle.cpp [hunk 22]
```diff
@@ -30,9 +30,9 @@ const AP_Param::GroupInfo AC_Circle::var_info[] = {
 // Note that the Vector/Matrix constructors already implicitly zero
 // their values.
 //
-AC_Circle::AC_Circle(const AP_InertialNav& inav, const AP_AHRS_View& ahrs, AC_PosControl& pos_control) :
-    _inav(inav),
+AC_Circle::AC_Circle(const AP_AHRS_NavEKF& ahrs, const AP_AHRS_View& ahrs_view, AC_PosControl& pos_control) :
```
Comment (OXINARF) [inline libraries/AC_WPNav/AC_Circle.cpp]: Same about ahrs and ahrs_view

File: ArduPlane/quadplane.cpp [hunk 23]
```diff
@@ -1705,7 +1705,9 @@ void QuadPlane::vtol_position_controller(void)
         
         pos_control->update_vel_controller_xyz(ekfNavVelGainScaler);
 
-        const Vector3f& curr_pos = inertial_nav.get_position();
+        Vector2f curr_pos;
```
Comment (khancyr) [inline ArduPlane/quadplane.cpp]: Vector2f is ctr fill with 0 by default

File: ArduPlane/quadplane.cpp [hunk 24]
```diff
@@ -749,7 +749,10 @@ void QuadPlane::run_z_controller(void)
         // starts the Z controller off with the right values
         gcs().send_text(MAV_SEVERITY_INFO, "Reset alt target to %.1f", (double)inertial_nav.get_altitude());
         pos_control->set_alt_target(inertial_nav.get_altitude());
-        pos_control->set_desired_velocity_z(inertial_nav.get_velocity_z());
+        Vector3f vel;
```
Comment (khancyr) [inline ArduPlane/quadplane.cpp]: Vector3f is ctr fill with 0 by default

File: ArduCopter/inertia.cpp [hunk 25]
```diff
@@ -17,6 +17,13 @@ void Copter::read_inertia()
     current_loc.lng = loc.lng;
     current_loc.lat = loc.lat;
 
+    // Get XYZ position and velocity in NEU and cm
```
Comment (khancyr) [inline ArduCopter/inertia.cpp]: What is wrong ? the XYZ part ?

File: libraries/AC_Fence/AC_Fence.cpp [hunk 26]
```diff
@@ -132,7 +131,12 @@ bool AC_Fence::pre_arm_check(const char* &fail_msg) const
     }
 
     // if we have horizontal limits enabled, check inertial nav position is ok
-    if ((_enabled_fences & (AC_FENCE_TYPE_CIRCLE | AC_FENCE_TYPE_POLYGON))>0 && !_inav.get_filter_status().flags.horiz_pos_abs && !_inav.get_filter_status().flags.pred_horiz_pos_abs) {
+    nav_filter_status filt_status;
+    if (!_ahrs.get_filter_status(filt_status)) {
+        fail_msg = "nav position unvailable";
```
Comment (khancyr) [inline libraries/AC_Fence/AC_Fence.cpp]: agree

File: ArduPlane/quadplane.cpp [hunk 27]
```diff
@@ -2143,13 +2152,15 @@ void QuadPlane::Log_Write_QControl_Tuning()
     const Vector3f &accel_target = pos_control->get_accel_target();
     Vector3f vel;
     ahrs.get_velocity_NED(vel);
+    float alt{};
+    ahrs.get_relative_position_D_origin(alt);
```
Comment (khancyr) [inline ArduPlane/quadplane.cpp]: agree, done

File: ArduPlane/quadplane.cpp [hunk 28]
```diff
@@ -2071,7 +2077,8 @@ void QuadPlane::check_land_complete(void)
         landing_detect.land_start_ms = 0;
         return;
     }
-    float height = inertial_nav.get_altitude()*0.01f;
+    float height{};
+    ahrs.get_relative_position_D_origin(height);
```
Comment (khancyr) [inline ArduPlane/quadplane.cpp]: agree, done

File: libraries/AC_Fence/AC_Fence.cpp [hunk 29]
```diff
@@ -277,8 +283,9 @@ bool AC_Fence::check_destination_within_fence(const Location_Class& loc)
     if ((get_enabled_fences() & AC_FENCE_TYPE_POLYGON) && _boundary_num_points > 0) {
         // check ekf has a good location
         Location temp_loc;
-        if (_inav.get_location(temp_loc)) {
-            const struct Location &ekf_origin = _inav.get_origin();
+        if (_ahrs.get_location(temp_loc)) {
+            struct Location ekf_origin {};
+            _ahrs.get_origin(ekf_origin);
             Vector2f position = location_diff(ekf_origin, loc) * 100.0f;
```
Comment (khancyr) [inline libraries/AC_Fence/AC_Fence.cpp]: yes, good catch

File: libraries/AC_Avoidance/AC_Avoid.cpp [hunk 30]
```diff
@@ -395,19 +400,20 @@ void AC_Avoid::limit_velocity(float kP, float accel_cmss, Vector2f &desired_vel,
...
 }
 
 /*
  * Gets the altitude above home in cm
  */
 float AC_Avoid::get_alt_above_home() const
 {
-    // vehicle's alt above ekf origin + ekf origin's alt above sea level - home's alt above sea level
-    return _inav.get_altitude() + _inav.get_origin().alt - _ahrs.get_home().alt;
+    // vehicle's alt above ekf origin - ekf origin's alt above sea level + home's alt above sea level
```
Comment (khancyr) [inline libraries/AC_Avoidance/AC_Avoid.cpp]: agree, done

File: libraries/AC_AttitudeControl/AC_PosControl.cpp [hunk 31]
```diff
@@ -299,9 +323,11 @@ void AC_PosControl::get_stopping_point_z(Vector3f& stopping_point) const
 /// init_takeoff - initialises target altitude if we are taking off
 void AC_PosControl::init_takeoff()
 {
-    const Vector3f& curr_pos = _inav.get_position();
+    Vector3f curr_pos;
+    _ahrs.get_relative_position_NED_origin(curr_pos);
```
Comment (khancyr) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp]: agee, but should get_relative_position_D_origin use baro alt is case of problem like get_relative_position_D_home ?

File: libraries/AC_AttitudeControl/AC_PosControl.h [hunk 32]
```diff
@@ -48,7 +47,7 @@ class AC_PosControl
 public:
 
     /// Constructor
-    AC_PosControl(const AP_AHRS_View& ahrs, const AP_InertialNav& inav,
+    AC_PosControl(AP_AHRS_NavEKF& ahrs, const AP_AHRS_View& ahrs_view,
```
Comment (khancyr) [inline libraries/AC_AttitudeControl/AC_PosControl.h]: @OXINARF you are right just one is need, I would said that the view isn't need as no rotation is applied...

File: libraries/AC_WPNav/AC_WPNav.cpp [hunk 33]
```diff
@@ -342,14 +350,14 @@ void AC_WPNav::update_loiter(float ekfGndSpdLimit, float ekfNavVelGainScaler)
...
-    const Vector3f& curr_vel = _inav.get_velocity();
+    Vector2f curr_vel = _ahrs.groundspeed_vector() * 100.0f;  // m/s to cm/s
     Vector3f stopping_point;
 
     // initialise position controller
     _pos_control.init_xy_controller();
 
     // initialise pos controller speed and acceleration
-    _pos_control.set_speed_xy(curr_vel.length());
+    _pos_control.set_speed_xy(curr_vel.length()); // BUG? the speed_xy was calculated with z speed ?
```
Comment (khancyr) [inline libraries/AC_WPNav/AC_WPNav.cpp]: it was a bug, I had confirmed with @rmackay9

**Comment by khancyr** [comment]
@OXINARF thanks for review, I update with error.
About Vector not explicitly initialized, their constructor init them at zero already, so I don't think it is needed to explicitly init to 0, what do you think ?
About ahrs_view, copter always init it with rotation_null so it should have difference with AHRS no ? If I am right, in most case we could remove ahrs_view for direct ahrs_nav

File: ArduCopter/inertia.cpp (hunk 25 above)
Comment (OXINARF) [inline ArduCopter/inertia.cpp]: No, actually the NEU and cm part. It gets it in NED and m and then does the conversions.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 31 above)
Comment (OXINARF) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp]: No, it shouldn't. Baro is reset every time you arm, which means it should have the same altitude above home, but not necessarily above origin. Home can also be changed and then baro wouldn't match it either, but it was done that way unfortunately.

File: libraries/AC_AttitudeControl/AC_PosControl.h (hunk 32 above)
Comment (OXINARF) [inline libraries/AC_AttitudeControl/AC_PosControl.h]: View is definitely needed as rotation is applied for quadplanes - that's why it was created. But I think that only the roll/pitch are rotated, not everything else, so we wouldn't need the rotated and not-rotated AHRS objects. We should get confirmation from @tridge.

**Comment by OXINARF** [comment]
> About Vector not explicitly initialized, their constructor init them at zero already, so I don't think it is needed to explicitly init to 0, what do you think ?

Yes, you are right, sorry. I forgot that the Vector classes had a default constructor setting to zero.

> About ahrs_view, copter always init it with rotation_null so it should have difference with AHRS no ? If I am right, in most case we could remove ahrs_view for direct ahrs_nav

I've answered directly in the comment. We do need ahrs_view because quadplane uses a rotation and that's why AHRS_View was created.

File: ArduCopter/inertia.cpp (hunk 25 above)
Comment (khancyr) [inline ArduCopter/inertia.cpp]: ok

File: libraries/AC_AttitudeControl/AC_PosControl.h (hunk 32 above)
Comment (khancyr) [inline libraries/AC_AttitudeControl/AC_PosControl.h]: ok I see, on case on quadplane use a rotation ... damned ! it would have simplified a lot !

Comment (OXINARF) [inline libraries/AC_AttitudeControl/AC_PosControl.h]: @khancyr I think you didn't understood my answer. We need ahrs_view, but I think we don't need ahrs - we just need to confirm that with @tridge.

Comment (khancyr) [inline libraries/AC_AttitudeControl/AC_PosControl.h]: Yes, I understand, it is just that ahrs_view is mostly a copying what is in ahrs, so if we could have remove ahrs_view, we could have simplied the codebase a little but that is not possible at least at one place the rotation from ahrs_view is used ! So we could do the other way and only keep ahrs_view ! But we will need to carefully review the constructor to not passing a ahrs_nav to ahrs_view ...

**Comment by amilcarlucas** [comment]
Travis was overloaded and hit the 50 minute limit on this one. Otherwise from that, what is blocking this one ?

**Comment by amilcarlucas** [comment]
We are using this for rover obstacle avoidance. It is easier to develop and test the algorithms in a Rover than on a copter.

We do need some more time to test it further on our rovers.
Is anyone testing this one on copters ?

BTW it needs a rebase now.

**Comment by amilcarlucas** [comment]
Sorry to annoy, but is this needed to get proximity and fence into Rover ?
If this if really, really required, why is it not further pursued ?

**Comment by khancyr** [comment]
@amilcarlucas IMO it is need for rover and other vehicle as inertial lib doesn't do much thing ... and it would be a waste to add it on rover to remove it soon after.

There are still some unanswered question about EKF and what append in case of problem. I will rebase that next week and try gather people to test (some place need to be looked when we lost position or got EKF problem) and merge it.

**Comment by amilcarlucas** [comment]
@khancyr No activity for 3 months :(

**Comment by khancyr** [comment]
Closed in favor of https://github.com/ArduPilot/ardupilot/pull/7345
